README.md
## chshli: Check Shared Libraries

### What Is This For?

Package management is a wonderful tool. If integrated with the operating system, it allows
us to install and upgrade components with great ease and reliability.  It is also pleasant
at the application level: CRAN functions really well thanks to `install.packages()` and
`update.packages()` even though these are generally unaware of the host OS (which can lead
to occassional frustrations).  The fact that the _OS_ package management system and the
_application_ package management system do not know about each other can lead to
occassional frictions.

There are two specific issues this package tries to help with.  We discuss both below.

#### Use Case 1: Distribution Upgrade

Say you just upgraded from Ubuntu 19.10 to Ubuntu 20.04, as I just did this week (in late
May of 2020).  You may have a number of CRAN packages installed from source in
`/usr/local/lib/R/site-library`.  All fine.

But `apt` and `dpkg` do not know about them. So it may remove libraries _not knowing about
these implicit dependencies_.  Next time you try to load one of the affected packages, it
will fail.

*Problem Illustration*

```r
R> library(git2r)
Error: package or namespace load failed for ‘git2r’ in dyn.load(file, DLLpath = DLLpath, ...):
 unable to load shared object '/usr/local/lib/R/site-library/git2r/libs/git2r.so':
  libgit2.so.27: cannot open shared object file: No such file or directory
R> 
```

The `chshli` package can help here:

*Example:*

```r
R> library(chshli)
Attaching chshli on ubuntu:20.04
R> checkSharedLibs(db="19.10")
Looking at libxml2.so.2 :       libxml2 
Looking at libapt-pkg.so.5.90 : libapt-pkg5.90 
Looking at libgit2.so.27 :      libgit2-27 
NULL
R>
```

This identifies `libgit2` (likely via package [git2r](https://cran.r-project.org/package=git2r)), 
`libapt-pkg`  (likely via [RcppApt](https://cran.r-project.org/package=RcppAPT) and 
`libxml2` (likely via [xml2](https://cran.r-project.org/package=xml2)). All three R packages 
can probably be fixed via a simple reinstallation from source.


#### Use Case 2: RSPM Installations

Say you experiment with RSPM, for example via [this Rocker
container](https://github.com/rocker-org/rocker/tree/master/r-rspm/bionic) which you can
pull via `docker pull rocker/r-rspm:18.04`.  And you did just run
`install.packages("xml2")`.  But maybe you cannot load it as the next example shows.

*Problem Illustration*

```r
> install.packages("xml2")
Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)
trying URL 'https://packagemanager.rstudio.com/all/__linux__/bionic/latest/src/contrib/xml2_1.3.2.tar.gz'
Content type 'binary/octet-stream' length 519621 bytes (507 KB)
==================================================
downloaded 507 KB

* installing *binary* package ‘xml2’ ...
* DONE (xml2)

The downloaded source packages are in
        ‘/tmp/RtmpW0G7Mr/downloaded_packages’
> library(xml2)
Error: package or namespace load failed for ‘xml2’ in dyn.load(file, DLLpath = DLLpath, ...):
 unable to load shared object '/usr/local/lib/R/site-library/xml2/libs/xml2.so':
  libxml2.so.2: cannot open shared object file: No such file or directory
> 
```

Here RSPM helps us with a quick and painless download of a binary. Which we then cannot load 
as it depends on an _OS-level_ package.  

The `chshli` package can help here:


*Example:*

```r
R> library(chshli)
Attaching chshli on ubuntu:18.04
R> checkSharedLibs()
Looking at libxml2.so.2 :       libxml2 
NULL
> 
```

and installing (at the system-level, _i.e._ as `root`) `sudo apt install libxml2` will make your
package work.  Which is a very nice complement to the nice and fast installation via RSPM.

(And we should note that we now add "a few" libraries to the aforementioned RSPM containers. So
`xml2` will already be covered if you try to replicate this.)


### Installation

For now from GitHub via `remotes::install_github("eddelbuettel/chshli")` or, if you have
[littler](https://github.com/eddelbuettel/littler), via `installGithub eddelbuettel/chshli`.

### Use with other Distributions

Note that you may need to provide the backend end file for your distribution. As these are large we
do not ship them. See `inst/rawdata/` and `inst/extdata` in the source, and try to follow what I did
for Ubuntu if your system is different.

### Author

Dirk Eddelbuettel 

### License
                                                                                                
GPL (>= 2) 


ChangeLog
2020-07-26  Dirk Eddelbuettel  <edd@debian.org>

	* R/libs.R: Minor tweaks to defaults and messages

	* inst/rawdata/20.04/: Added
	* inst/extdata/libContents-amd64.20.04.gz: Idem

2020-05-28  Dirk Eddelbuettel  <edd@debian.org>

	* README.md: Expanded description

	* NAMESPACE: Add read.table imports

	* R/libs.R (checkSharedLibs): Add documentation
	* man/checkSharedLibs.Rd: Idem

2020-05-27  Dirk Eddelbuettel  <edd@debian.org>

	* DESCRIPTION: Initial version 0.0.1



R/libs.R
.libs <- function() {
    if (!.pkgenv$goodsystem)
        return(invisible(NULL))

    if (Sys.which("ldd") == "")
        stop("No 'ldd'.", call.=FALSE)

    if (Sys.which("xargs") == "")
        stop("No 'xargs'.", call.=FALSE)

    ## we take find, grep, sed, ... as given

    .findLib <- function(path) {
        ##file <- system.file("libs", paste0(path, ".so"), package=path)
        ##return(if (nchar(file) > 0) file else NULL)
        readLines(pipe(sprintf("find %s -name \\*so -type f -size +1024c | xargs -r ldd | grep 'not found' | sed -e 's/ => not found//' -e's/\t//'", path)))
    }
    rl <- lapply(.libPaths(), .findLib)
    vec <- do.call(rbind, rl)
    unique(vec[order(vec)])
}

.getDB <- function(rel=c("18.04", "19.10"), arch="amd64") {
    rel <- match.arg(rel)
    db <- system.file("extdata", paste0("libContents-", arch, ".", rel, ".gz"), package="chshli")
    if (nchar(db) == 0)
        stop("No data for '", rel, "' on '", arch,
             "'. See the 'rawdata' and 'extdata' directories in the sources.", call.=FALSE)
    dat <- read.table(db, row.names=1, col.names=c("path", "pkg"))
    dat
}

##' Check shared libraries on current system
##'
##' This function examine the shared libraries in argument
##' \code{vec}. It runs \code{ldd} on these, filters for \dQuote{not
##' found} to identify unresolved shared libraries and tries to map
##' these against (stored) information on distribution packages (where
##' Ubuntu 18.04 and 19.10 are currently supported).
##'
##' The internal helper function \code{.libs()} can be used to
##' identify shared libraries from packages in the current
##' \code{.libPaths()}.  Because search all packages in each 
##' \code{.libPaths()} entry at once (using shell tools) it does
##' not currently associate finds with the packages they originate
##' from.
##'
##' @title Check Shared Libraries
##' @param vec A (optional) character vector of shared libraries,
##'     typcially with major soname. If missing a default vector is
##'     used as fallback.
##' @param db An optional identifier for a library package database,
##'     should be one of "18.04" or "19.10".
##' @return Nothing currently but information is printed.
##' @author Dirk Eddelbuettel
##' @examples 
##' checkSharedLibs(c("libxml2.so.2"))
checkSharedLibs <- function(vec, db) {
    if (missing(vec)) {
        ## mostly for debugging / dev, to be removed eventually
        ##vec <- c("libxml2.so.2", "libapt-pkg.so.5.90", "libgit2.so.27", "libqdb_api.so")
        message("Setting default 'vec' argument based on .libPaths() scan.")
        vec <- .libs()
    }

    if (missing(db))
        db <- .getDB()
    else
        db <- .getDB(db)

    for (v in vec) {
        cat("Looking at", v, ":\t")
        ind <- grep(paste0("/", v), rownames(db))
        if (length(ind) > 0)
            cat(gsub(".*/", "", unique(db[ind, "pkg"])), "\n")
        else
            cat("<NA>\n")
    }
    invisible()
}
